scripts/convert_ima_policy.py

```diff
@@ -0,0 +1,173 @@
+import argparse
+import copy
+import datetime
+import json
+import re
+import sys
+from os.path import basename
+
+from keylime.ima import file_signatures, ima
+
+"""
+This script converts legacy Keylime policies into the unified, next-generation
+IMA policy format.Both legacy allow/exclude lists as well as out-of-date IMA
+policy formats are accepted.
+
+Example usage:
+
+To convert an allowlist and exclude list to an IMA policy at `keylime-policy.json`:
+
+```
+python convert_ima_policy.py -a <allowlist path> -e <exclude list path> -o keylime-policy.json
+```
+
+To convert an existing IMA policy to the latest version:
+
+```
+python convert_ima_policy.py -i <ima policy path> -o keylime-policy.json
+```
+
+To view help and see all available options:
+
+```
+python convert_ima_policy.py -h
+```
+"""
+
+# Creates an IMA policy from provided legacy allowlist.
+def create_ima_policy(allowlist_path):
+    with open(allowlist_path, "r") as f:
+        alist_raw = f.read()
+
+    p = re.compile(r"^\s*{")
+    if p.match(alist_raw):
+        print(f"{basename(allowlist_path)} is JSON-formatted; converting to IMA policy")
+        alist_json = json.loads(alist_raw)
+        ima_policy = copy.deepcopy(ima.EMPTY_IMA_POLICY)
+        ima_policy["meta"]["timestamp"] = str(datetime.datetime.now())
+        ima_policy["meta"]["generator"] = "convert-ima-policy"
+        for key in alist_json.keys():
+            if key == "hashes":
+                ima_policy["digests"] = alist_json[key]
+            elif key == "meta":
+                continue
+            else:
+                ima_policy[key] = alist_json[key]
+    else:
+        print(f"{basename(allowlist_path)} is not JSON-formatted; converting to IMA policy from flat file format")
+        ima_policy = convert_legacy_allowlist(alist_raw)
+    return ima_policy
+
+
+# Updates an existing IMA policy to the latest version, and adds any provided input
+def update_ima_policy(policy, excludelist_path, verification_keys):
+    if policy["meta"]["version"] < ima.IMA_POLICY_CURRENT_VERSION:
+        print(
+            f"Provided policy has version {policy['meta']['version']}; latest policy has version {ima.IMA_POLICY_CURRENT_VERSION}. Updating to latest version."
+        )
+        updated_policy = copy.deepcopy(ima.IMA_POLICY_CURRENT_VERSION)
+        updated_policy["meta"]["timestamp"] = str(datetime.datetime.now())
+        updated_policy["meta"]["generator"] = "convert-ima-policy"
```

Comment:
This should probably just be the script name instead of a something that's kinda like it.